sa/storage-authority.go

```diff
@@ -237,6 +237,46 @@ func (ssa *SQLStorageAuthority) GetLatestValidAuthorization(registrationId int64
 	return ssa.GetAuthorization(auth.ID)
 }
 
+type TooManyCertificatesError string
+
+func (t TooManyCertificatesError) Error() string {
+	return string(t)
+}
+
+// CountCertificates returns the number of certificates issued within a time
+// period containing DNSNames that are equal to, or subdomains of, the given name.
+func (ssa *SQLStorageAuthority) CountCertificatesByName(domain string, earliest, latest time.Time) (int, error) {
+	var count int64
+	const max = 10000
+	var serials []struct {
+		Serial string
+	}
+	_, err := ssa.dbMap.Select(
+		&serials,
+		`SELECT serial from issuedNames
+		 WHERE (reversedName = :reversedDomain OR
+			      reversedName LIKE CONCAT(:reversedDomain, ".%"))
+		 AND Issued > :earliest AND Issued <= :latest
+		 LIMIT :limit;`,
+		map[string]interface{}{
+			"reversedDomain": core.ReverseName(domain),
+			"earliest":       earliest,
+			"latest":         latest,
+			"limit":          max + 1,
+		})
+	if err != nil {
+		return -1, err
+	} else if count > max {
+		return -1, TooManyCertificatesError(fmt.Sprintf("More than %d issuedName entries for %s.", max, domain))
+	}
+	serialMap := make(map[string]struct{}, len(serials))
+	for _, s := range serials {
+		serialMap[s.Serial] = struct{}{}
+	}
+
+	return len(serialMap), nil
```

Comment:
One certificate can have many names that are subdomains of the same parent, which means that the SELECT query above can have repeats of the same serial. For instance, imagine a certificate for www.example.com and example.com.

We could solve this with a simple DISTINCT, but @jmhodges pointed out that when the list of serials is in the thousands, as it easily could be, DISTINCT chews up more memory on the DB instance than we would like, and it's cheaper to do the distinct in application logic instead. That is, it's easier to run additional SAs than additional DBs.